From 2030319cfa0cacb3397fccd8e1cb0a16789b1676 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Wed, 8 Jun 2022 09:30:34 +0000
Subject: [PATCH] xen/arm: Add dummy xl-psr related domctl and sysctl for Arm

xl-psr provides a set of domctl and sysctl ops and commands that
can be reused by MPAM to get the MPAM hardware information, configure
a domain and read cache/membw monitor values. Therefore, this commit
implements the ops needed for now, namely XEN_DOMCTL_psr_cmt_op,
XEN_DOMCTL_psr_alloc, XEN_SYSCTL_psr_cmt_op and XEN_SYSCTL_psr_alloc,
as dummy domctl and sysctl to provide basic framework for xl-psr on
Arm. Following commits will fullfill these ops by calling related MPAM
helpers in these ops.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-3763
---
 xen/arch/arm/arm64/domctl.c | 13 +++++++++++++
 xen/arch/arm/sysctl.c       | 15 ++++++++++++++-
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/xen/arch/arm/arm64/domctl.c b/xen/arch/arm/arm64/domctl.c
index 14fc622e9956..56d993198698 100644
--- a/xen/arch/arm/arm64/domctl.c
+++ b/xen/arch/arm/arm64/domctl.c
@@ -12,6 +12,7 @@
 #include <public/domctl.h>
 #include <asm/arm64/sve.h>
 #include <asm/cpufeature.h>
+#include <asm/mpam.h>
 
 static long switch_mode(struct domain *d, enum domain_type type)
 {
@@ -55,6 +56,18 @@ long subarch_do_domctl(struct xen_domctl *domctl, struct domain *d,
         }
         break;
 
+#ifdef CONFIG_MPAM
+    case XEN_DOMCTL_psr_cmt_op:
+        if ( !cpu_has_mpam )
+            return -EOPNOTSUPP;
+        return 0;
+
+    case XEN_DOMCTL_psr_alloc:
+        if ( !cpu_has_mpam )
+            return -EOPNOTSUPP;
+        return 0;
+#endif
+
     default:
         return -ENOSYS;
     }
diff --git a/xen/arch/arm/sysctl.c b/xen/arch/arm/sysctl.c
index 5cda0dc674f6..b441f6a80eb3 100644
--- a/xen/arch/arm/sysctl.c
+++ b/xen/arch/arm/sysctl.c
@@ -14,6 +14,7 @@
 #include <xen/hypercall.h>
 #include <asm/arm64/sve.h>
 #include <public/sysctl.h>
+#include <asm/mpam.h>
 
 void arch_do_physinfo(struct xen_sysctl_physinfo *pi)
 {
@@ -26,7 +27,7 @@ void arch_do_physinfo(struct xen_sysctl_physinfo *pi)
 long arch_do_sysctl(struct xen_sysctl *sysctl,
                     XEN_GUEST_HANDLE_PARAM(xen_sysctl_t) u_sysctl)
 {
-    long ret;
+    long ret = 0;
 
     switch ( sysctl->cmd )
     {
@@ -34,6 +35,18 @@ long arch_do_sysctl(struct xen_sysctl *sysctl,
         ret = dt_overlay_sysctl(&sysctl->u.dt_overlay);
         break;
 
+#ifdef CONFIG_MPAM
+    case XEN_SYSCTL_psr_cmt_op:
+        if ( !cpu_has_mpam )
+            return -EOPNOTSUPP;
+        break;
+
+    case XEN_SYSCTL_psr_alloc:
+        if ( !cpu_has_mpam )
+            return -EOPNOTSUPP;
+        break;
+#endif
+
     default:
         ret = -ENOSYS;
         break;
-- 
2.34.1

